cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/functional/BasicLattices.kt

```diff
@@ -0,0 +1,251 @@
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.helpers.functional
+
+import de.fraunhofer.aisec.cpg.helpers.IdentitySet
+import de.fraunhofer.aisec.cpg.helpers.identitySetOf
+import de.fraunhofer.aisec.cpg.helpers.toIdentitySet
+import java.util.IdentityHashMap
+import kotlin.Pair
+import kotlin.collections.component1
+import kotlin.collections.component2
+import kotlin.collections.toMap
+
+/**
+ * A complete lattice is an ordered structure of values of type [T]. [T] could be anything, e.g., a
+ * set, a new data structure (like a range), or anything else. [T] depends on the analysis and
+ * typically has to abstract the value for the specific purpose.
+ *
+ * This class is actually used to hold individual instances of the lattice's elements and to compute
+ * bigger elements depending on these two elements.
+ *
+ * Implementations of this class have to implement the comparator, the least upper bound of two
+ * lattices.
+ */
+interface LatticeElement<T> : Comparable<LatticeElement<T>> {
+    val elements: T
+
+    /**
+     * Computes the least upper bound of this lattice and [other]. It returns a new object and does
+     * not modify either of the objects.
+     */
+    fun lub(other: LatticeElement<out T>): LatticeElement<in T>
+
+    /** Duplicates the object, i.e., makes a deep copy. */
+    fun duplicate(): LatticeElement<in T>
+}
+
+typealias PowersetLatticeT<V> = PowersetLattice<IdentitySet<V>, V>
+
+inline fun <reified V> emptyPowersetLattice() = PowersetLattice<IdentitySet<V>, V>(identitySetOf())
+
+/**
+ * Implements the [LatticeElement] for a lattice over a set of nodes. The lattice itself is
+ * constructed by the powerset.
+ */
+open class PowersetLattice<V : IdentitySet<T>, T>(override val elements: V) : LatticeElement<V> {
+    override fun lub(other: LatticeElement<out V>): PowersetLattice<in V, T> {
+        val newElements = this.elements.toIdentitySet()
+        newElements += other.elements
+        return PowersetLattice(newElements)
+    }
+
+    override fun duplicate() = PowersetLattice(this.elements.toIdentitySet() as V)
+
+    override fun compareTo(other: LatticeElement<V>): Int {
+        return if (this.elements == other.elements) {
+            0
+        } else if (this.elements.containsAll(other.elements)) {
+            1
+        } else {
+            -1
+        }
+    }
+
+    override fun equals(other: Any?): Boolean {
+        // The call of `toSet` ensures that we don't get stuck for different types of sets.
+        return other is PowersetLattice<V, T> && this.elements.toSet() == other.elements.toSet()
```

Comment:
Is there no other way? this does not seem like a good idea

Reply:
Since the value of both elements should be a `IdentitySet` now, this should not be required and I removed it